packages/protobuf/src/private/feature-set.ts

```diff
@@ -0,0 +1,122 @@
+
+/**
+ * Static edition feature defaults supported by @bufbuild/protobuf.
+ */
+export const featureSetDefaults = FeatureSetDefaults.fromBinary(
+  protoBase64.dec(
+    /*upstream-inject-feature-defaults-start*/ "ChESDAgBEAIYAiABKAEwAhjmBwoREgwIAhABGAEgAigBMAEY5wcKERIMCAEQARgBIAIoATABGOgHIOYHKOgH" /*upstream-inject-feature-defaults-end*/,
+  ),
+);
+
+/**
+ * A merged google.protobuf.FeaturesSet, with all fields guaranteed to be set.
+ */
+export type MergedFeatureSet = FeatureSet & Required<FeatureSet>;
+
+/**
+ * A function that resolves features for the given edition.
+ *
+ * If no feature set is provided, the default feature set for the edition is
+ * returned. If features are provided, they are merged into the edition default
+ * features.
+ */
+export type FeatureResolverFn = (
+  edition: Edition,
+  a?: FeatureSet,
+  b?: FeatureSet,
+) => MergedFeatureSet;
+
+/**
+ * Create an edition feature resolver with the given feature set defaults.
+ */
+export function createFeatureResolver(
+  compiledFeatureSetDefaults: FeatureSetDefaults,
+): FeatureResolverFn {
+  const min = compiledFeatureSetDefaults.minimumEdition ?? 0;
+  const max = compiledFeatureSetDefaults.maximumEdition ?? 0;
+  const defaultsBinByEdition = new Map<Edition, Uint8Array>();
+  return (edition: Edition, ...rest): MergedFeatureSet => {
+    let defaultsBin: Uint8Array | undefined = defaultsBinByEdition.get(edition);
+    if (defaultsBin === undefined) {
+      if (edition < min) {
+        throw new Error(
+          `Edition ${Edition[edition]} is earlier than the minimum supported edition ${Edition[min]}`,
+        );
+      }
+      if (max < edition) {
+        throw new Error(
+          `Edition ${Edition[edition]} is later than the maximum supported edition ${Edition[max]}`,
+        );
+      }
+      let highestMatch: { e: Edition; f: FeatureSet } | undefined = undefined;
+      for (const c of compiledFeatureSetDefaults.defaults) {
+        const e = c.edition ?? 0;
+        if (e > edition) {
+          continue;
+        }
+        if (highestMatch !== undefined && highestMatch.e > e) {
+          continue;
+        }
+        highestMatch = {
+          e,
+          f: c.features ?? new FeatureSet(),
+        };
+      }
+      if (highestMatch === undefined) {
+        throw new Error(
+          `No valid default found for edition ${Edition[edition]}`,
+        );
+      }
+      defaultsBin = highestMatch.f.toBinary();
+      defaultsBinByEdition.set(edition, defaultsBin);
+    }
+    const f = FeatureSet.fromBinary(defaultsBin);
+    for (const c of rest) {
+      if (c !== undefined) {
+        f.fromBinary(c.toBinary());
+      }
+    }
+    if (!validateMergedFeatures(f)) {
+      throw new Error(`Invalid FeatureSet for edition ${Edition[edition]}`);
+    }
+    return f;
+  };
+}
+
+function validateMergedFeatures(
+  featureSet: FeatureSet,
+): featureSet is MergedFeatureSet {
+  for (const p of [
```

Comment:
So if we pair this set of fixed properties with a newer `FeatureSet` definition, we'll end up violating the type instead of this being an error, right?

Perhaps better would be to a constant literal of type `MergedFeatureSet`, so the compiler can verify it is correct (e.g. has all required properties), and then here loop over the keys of the constant.

Reply:
Thanks for raising this question, I was meaning to take a closer look at forward compatibility. In https://github.com/bufbuild/protobuf-es/pull/627/commits/169354c25b094ce370993ae6cc5a3457ce10590f, I'm adding the following comments to this function:

> // When protoc generates google.protobuf.FeatureSetDefaults, it ensures that
// fields are not repeated or required, do not use oneof, and have a default
// value.
//
// When features for an element are resolved, features of the element and its
// parents are merged into the default FeatureSet for the edition. Because unset
// fields in the FeatureSet of an element do not unset the default FeatureSet
// values, a resolved FeatureSet is guaranteed to have all fields set. This is
// also the case for extensions to FeatureSet that a user might provide, and for
// features from the future.
//
// We cannot exhaustively validate correctness of FeatureSetDefaults at runtime
// without knowing the schema: If no value for a feature is provided, we do not
// know that it exists at all.
//
// As a sanity check, we validate that all fields known to our version of
// FeatureSet are set.

The gist is that I don't see an alternative to trusting the FeatureSetDefaults to be valid. If the defaults are valid, feature resolution will always yield a message with all fields set. 

In https://github.com/bufbuild/protobuf-es/pull/627/commits/ad1965dc2ffff83e845e108520289a70032489d8, I've updated the logic to use the field information embedded in the generated code to check fields. So this sanity check will always include all known fields.